engine/datahistory_manager.go

```diff
@@ -0,0 +1,857 @@
+		defer func() {
+			err = m.Stop()
+			if err != nil {
+				log.Error(log.DataHistory, err)
+			}
+		}()
+		return nil, fmt.Errorf("error retrieving jobs, has everything been setup? Data history manager will shut down. %w", err)
+	}
+	err = m.compareJobsToData(jobs...)
+	if err != nil {
+		return nil, err
+	}
+
+	return jobs, nil
+}
+
+func (m *DataHistoryManager) compareJobsToData(jobs ...*DataHistoryJob) error {
+	if m == nil {
+		return ErrNilSubsystem
+	}
+	if atomic.LoadInt32(&m.started) == 0 {
+		return ErrSubSystemNotStarted
+	}
+	var err error
+	for i := range jobs {
+		jobs[i].rangeHolder, err = kline.CalculateCandleDateRanges(jobs[i].StartDate, jobs[i].EndDate, jobs[i].Interval, uint32(jobs[i].RequestSizeLimit))
+		if err != nil {
+			return err
+		}
+		var candles kline.Item
+		switch jobs[i].DataType {
+		case dataHistoryCandleDataType:
+			candles, err = m.candleLoader(jobs[i].Exchange, jobs[i].Pair, jobs[i].Asset, jobs[i].Interval, jobs[i].StartDate, jobs[i].EndDate)
+			if err != nil && !errors.Is(err, candle.ErrNoCandleDataFound) {
+				return err
+			}
+			jobs[i].rangeHolder.SetHasDataFromCandles(candles.Candles)
+		case dataHistoryTradeDataType:
+			err := m.tradeLoader(jobs[i].Exchange, jobs[i].Asset.String(), jobs[i].Pair.Base.String(), jobs[i].Pair.Quote.String(), jobs[i].rangeHolder)
+			if err != nil && !errors.Is(err, sql.ErrNoRows) {
+				return err
+			}
+		default:
+			return errUnknownDataType
+		}
+	}
+	return nil
+}
+
+func (m *DataHistoryManager) run() {
+	go func() {
+		validJobs, err := m.PrepareJobs()
+		if err != nil {
+			log.Error(log.DataHistory, err)
+		}
+		m.m.Lock()
+		m.jobs = validJobs
+		m.m.Unlock()
+
+		for {
+			select {
+			case <-m.shutdown:
+				return
+			case <-m.interval.C:
+				if m.databaseConnectionInstance.IsConnected() {
+					go func() {
+						if err := m.runJobs(); err != nil {
+							log.Error(log.DataHistory, err)
+						}
+					}()
+				}
+			}
+		}
+	}()
+}
+
+func (m *DataHistoryManager) runJobs() error {
+	if m == nil {
+		return ErrNilSubsystem
+	}
+	if atomic.LoadInt32(&m.started) == 0 {
+		return ErrSubSystemNotStarted
+	}
+
+	if !atomic.CompareAndSwapInt32(&m.processing, 0, 1) {
+		return fmt.Errorf("runJobs %w", errAlreadyRunning)
+	}
+	defer atomic.StoreInt32(&m.processing, 0)
+
+	validJobs, err := m.PrepareJobs()
+	if err != nil {
+		return err
+	}
+
+	m.m.Lock()
+	defer func() {
+		m.m.Unlock()
+	}()
+	m.jobs = validJobs
+	log.Infof(log.DataHistory, "processing data history jobs")
+	for i := 0; (i < int(m.maxJobsPerCycle) || m.maxJobsPerCycle == -1) && i < len(m.jobs); i++ {
+		err := m.runJob(m.jobs[i])
+		if err != nil {
+			log.Error(log.DataHistory, err)
+		}
+		if m.verbose {
+			log.Debugf(log.DataHistory, "completed run of data history job %v", m.jobs[i].Nickname)
+		}
+	}
+	log.Infof(log.DataHistory, "completed run of data history jobs")
+
+	return nil
+}
+
+// runJob will iterate
+func (m *DataHistoryManager) runJob(job *DataHistoryJob) error {
```

Comment:
Something funky potentially going on in here as discussed which prevents the job status of a large job which might have missing bits of data to be completed and status to be changed accordingly.

Reply:
I've just pushed through an update which will allow your jobs to finish. When you check it out, let me know your thoughts on how it executes